### PR TITLE
ARROW-9299: [C++][Python] Expose ORC metadata

### PR DESCRIPTION
**Pull request by iajoiner**

### CODE REVIEW

**Comment by mathyingzhou** [comment]
@pitrou Please review. :)

**Comment by calebwin** [comment]
Hi @mathyingzhou,

Thanks for working on this and this looks great. I think the main things we wanted were schema, stripe-level statistics, and even row-level statistics (lower priority than stripe-level).

**Comment by mathyingzhou** [comment]
@calebwin Really thanks! For schema there is already ORCFile.schema(). For the statistics we will need to add the adapter to some other ORC functions which will take a while. I suddenly got very busy at work so I will need 4 extra weeks before this can be delivered. Sorry for the inconvenience..

File: cpp/src/arrow/adapters/orc/adapter.cc [hunk 1]
```diff
@@ -229,6 +229,17 @@ class ORCFileReader::Impl {
     return GetArrowSchema(type, out);
   }
 
+  Result<std::shared_ptr<KeyValueMetadata>> ReadMetadata() {
+    std::list<std::string> keys = reader_->getMetadataKeys();
```
Comment (pitrou) [inline cpp/src/arrow/adapters/orc/adapter.cc:233]: Make this `const`?

File: cpp/src/arrow/adapters/orc/adapter.cc [hunk 2]
```diff
@@ -229,6 +229,17 @@ class ORCFileReader::Impl {
...
+    auto metadata = std::make_shared<KeyValueMetadata>();
+    if (!keys.empty()) {
```
Comment (pitrou) [inline cpp/src/arrow/adapters/orc/adapter.cc:235]: Why not `for (const auto& key : keys) {`?

File: cpp/src/arrow/adapters/orc/adapter.h [hunk 3]
```diff
@@ -27,6 +27,7 @@
 #include "arrow/status.h"
 #include "arrow/type.h"
 #include "arrow/type_fwd.h"
+#include "arrow/util/key_value_metadata.h"
```
Comment (pitrou) [inline cpp/src/arrow/adapters/orc/adapter.h:30]: No need for this a priori, as `KeyValueMetadata` is already declared in `arrow/type_fwd.h`.

File: cpp/src/arrow/adapters/orc/adapter.h [hunk 4]
```diff
@@ -149,7 +156,7 @@ class ARROW_EXPORT ORCFileWriter {
   /// \brief Creates a new ORC writer.
   ///
   /// \param[in] output_stream a pointer to the io::OutputStream to write into
-  /// \return the returned writer object
+  /// \return the returned writer object or Status if error occurs
```
Comment (pitrou) [inline cpp/src/arrow/adapters/orc/adapter.h:159]: There is no need to state this explicitly, as `Result<>` can implicitly convey an error.

File: cpp/src/arrow/adapters/orc/adapter_test.cc [hunk 5]
```diff
@@ -289,6 +289,8 @@ TEST(TestAdapterRead, ReadIntAndStringFileMultipleStripes) {
   constexpr uint64_t reader_batch_size = 1024;
 
   auto writer = CreateWriter(stripe_size, *type, &mem_stream);
+  // writer->addUserMetadata("key1", "value1");
+  // writer->addUserMetadata("key2", "value2");
```
Comment (pitrou) [inline cpp/src/arrow/adapters/orc/adapter_test.cc:293]: Why the commented out code?

File: python/pyarrow/tests/test_orc.py [hunk 6]
```diff
@@ -110,20 +114,20 @@ def check_example_file(orc_path, expected_df, need_fix=False):
...
-        check_example_values(pd.DataFrame(batch.to_pydict()),
-                             expected_df,
-                             start=json_pos,
-                             stop=json_pos + len(batch))
+        check_example_values(
+            pd.DataFrame(batch.to_pydict()),
+            expected_df,
+            start=json_pos,
+            stop=json_pos + len(batch),
+        )
```
Comment (pitrou) [inline python/pyarrow/tests/test_orc.py:122]: Please avoid doing unrelated style changes.
Also, if you want an automated style pass, you can use `archery lint --python --fix`
(see https://arrow.apache.org/docs/developers/archery.html)

**Review by pitrou — CHANGES_REQUESTED** [review]
I'm not sure how useful this is since you can already extract the metadata from the schema, but I guess it doesn't hurt...

File: cpp/src/arrow/adapters/orc/adapter.cc (hunk 2 above)
Comment (mathyingzhou) [reply cpp/src/arrow/adapters/orc/adapter.cc:235]: Sure!

File: cpp/src/arrow/adapters/orc/adapter.cc (hunk 1 above)
Comment (mathyingzhou) [reply cpp/src/arrow/adapters/orc/adapter.cc:233]: Yup.

File: cpp/src/arrow/adapters/orc/adapter.h (hunk 3 above)
Comment (mathyingzhou) [reply cpp/src/arrow/adapters/orc/adapter.h:30]: Ah I see..

**Comment by mathyingzhou** [comment]
@pitrou Done! Please review again.

File: python/pyarrow/includes/common.pxd [hunk 7]
```diff
@@ -34,6 +34,7 @@ cimport cpython
 
 cdef extern from * namespace "std" nogil:
     cdef shared_ptr[T] static_pointer_cast[T, U](shared_ptr[U])
+    cdef shared_ptr[T] const_pointer_cast[T, U](shared_ptr[U])
```
Comment (pitrou) [inline python/pyarrow/includes/common.pxd:37]: This doesn't seem used?

**Review by pitrou — CHANGES_REQUESTED** [review]
Thanks for the update. Some more comments below.

File: python/pyarrow/orc.py [hunk 8]
```diff
@@ -26,11 +26,13 @@
...
-            typ.value_type.num_fields == 2 and
-            typ.value_type[0].name == 'key' and
-            typ.value_type[1].name == 'value')
+    return (
+        types.is_list(typ) and
+        types.is_struct(typ.value_type) and
+        typ.value_type.num_fields == 2 and
+        typ.value_type[0].name == "key" and
+        typ.value_type[1].name == "value"
+    )
```
[the remark this answers is not in the document]
Comment (mathyingzhou) [reply python/pyarrow/orc.py:35]: Sure..

File: python/pyarrow/_orc.pyx [hunk 9]
```diff
@@ -22,16 +22,19 @@
 from cython.operator cimport dereference as deref
 from libcpp.vector cimport vector as std_vector
 from libcpp.utility cimport move
+from libcpp.memory cimport const_pointer_cast
```
[the remark this answers is not in the document]
Comment (mathyingzhou) [reply python/pyarrow/_orc.pyx:25]: Ah sorry I forgot to remove it.

**Comment by mathyingzhou** [comment]
@pitrou Thanks for the comments! Please review again. None of the failures has anything to do with my PR.

**Review by pitrou — APPROVED** [review]
+1, thanks for the update!
